File: UDF Dönüştürücü/Core/Tools/UDFStructureWriter.swift
import Foundation

/// Yapılandırılmış UDF modelini UYAP content.xml (CDATA + elements + headers + footers) biçimine yazar.
enum UDFStructureWriter {

    struct BuildResult {
        let cdataText: String
        let elementsXML: String
        let headersXML: String?
        let footersXML: String?
    }

    static func build(from document: UDFEditDocument) -> BuildResult {
        var cdata = ""
        var offset = 0

        let headersXML = writeHeaderFooterSection(
            document.headers,
            itemTag: "header",
            cdata: &cdata,
            offset: &offset
        )

        var elements = "\n"
        for block in document.blocks {
            switch block {
            case .paragraph(let para):
                elements += writeParagraph(para, cdata: &cdata, offset: &offset)
            case .table(let table):
                elements += writeTable(table, cdata: &cdata, offset: &offset)
            }
        }

        let footersXML = writeHeaderFooterSection(
            document.footers,
            itemTag: "footer",
            cdata: &cdata,
            offset: &offset
        )

        return BuildResult(
            cdataText: cdata,
            elementsXML: elements,
            headersXML: headersXML,
            footersXML: footersXML
        )
    }

    // MARK: - Header / Footer

    private static func writeHeaderFooterSection(
        _ items: [UDFEditHeaderFooter],
        itemTag: String,
        cdata: inout String,
        offset: inout Int
    ) -> String? {
        guard !items.isEmpty else { return nil }
        var xml = "\n"
        for item in items {
            xml += "  <\(itemTag) type=\"\(escape(item.type))\">\n"
            for para in item.paragraphs {
                xml += writeParagraph(para, cdata: &cdata, offset: &offset, indent: "    ")
            }
            xml += "  </\(itemTag)>\n"
        }
        return xml
    }

    // MARK: - Paragraph

    private static func writeParagraph(_ para: UDFEditParagraph, cdata: inout String, offset: inout Int, indent: String = "") -> String {
        var tabAttr = ""
        if !para.tabStops.isEmpty, let first = para.tabStops.first {
            tabAttr = " TabSet=\"\(first):0:0\""
        }

        var xml = "\(indent)<paragraph SpaceAbove=\"\(para.spaceAbove)\" SpaceBelow=\"\(para.spaceBelow)\" "
            + "LeftIndent=\"\(para.leftIndent)\" RightIndent=\"\(para.rightIndent)\" "
            + "LineSpacing=\"\(para.lineSpacing)\" resolver=\"hvl-default\" "
            + "Alignment=\"\(para.alignment)\" Hanging=\"\(para.hangingIndent)\"\(tabAttr)>"

        let paraText = para.runs.map(\.text).joined()
        var runOffset = offset

        if paraText.isEmpty {
            xml += "<content resolver=\"hvl-default\" startOffset=\"\(offset)\" length=\"0\" />"
            cdata += "\n"
            offset += 1
        } else {
            for run in para.runs where !run.text.isEmpty {
                xml += runTag(run, startOffset: runOffset, length: run.text.count)
                runOffset += run.text.count
            }
            cdata += paraText + "\n"
            offset += paraText.count + 1
        }

        xml += "</paragraph>\n"
        return xml
    }

    // MARK: - Table

    private static func writeTable(_ table: UDFEditTable, cdata: inout String, offset: inout Int) -> String {
        let spans = table.columnSpans.map(String.init).joined(separator: ",")
        var spansAttr = spans.isEmpty ? "" : " columnSpans=\"\(spans)\""
        var xml = "<table tableName=\"Tablo\" columnCount=\"\(table.columnCount)\"\(spansAttr) border=\"\(escape(table.border))\">\n"

        for row in table.rows {
            xml += "  <row rowName=\"row\" rowType=\"\(escape(row.rowType))\" border=\"\(escape(table.border))\">\n"
            for cell in row.cells {
                var cellAttrs = "colspan=\"\(cell.colspan)\" rowspan=\"\(cell.rowspan)\" align=\"top\" border=\"\(escape(table.border))\" borderSpec=\"15\""
                if let fill = cell.fillColorARGB {
                    cellAttrs += " fillColor=\"\(fill)\""
                }
                xml += "    <cell \(cellAttrs)>\n"
                for para in cell.paragraphs {
                    xml += writeParagraph(para, cdata: &cdata, offset: &offset, indent: "      ")
                }
                xml += "    </cell>\n"
            }
            xml += "  </row>\n"
        }

        xml += "</table>\n"
        return xml
    }

    // MARK: - Run tag

    private static func runTag(_ run: UDFEditRun, startOffset: Int, length: Int) -> String {
        let tagName: String
        switch run.kind {
        case .content: tagName = "content"
        case .field: tagName = "field"
        case .space: tagName = "space"
        }

        var attrs = " resolver=\"hvl-default\""
        if run.isBold { attrs += " bold=\"true\"" }
        if run.isItalic { attrs += " italic=\"true\"" }
        if run.isUnderline { attrs += " underline=\"true\"" }
        if run.fontSize != 12 && run.fontSize > 0 {
            attrs += " size=\"\(Int(run.fontSize))\""
        }
        attrs += " family=\"\(escape(run.fontFamily))\""
        if let fg = run.foregroundARGB {
            attrs += " foreground=\"\(fg)\""
        }
        if let bg = run.backgroundARGB {
            attrs += " background=\"\(bg)\""
        }
        if case .field(let name) = run.kind {
            attrs += " fieldName=\"\(escape(name))\""
        }
        attrs += " startOffset=\"\(startOffset)\" length=\"\(length)\""
        return "<\(tagName)\(attrs) />"
    }

    private static func escape(_ string: String) -> String {
        string.replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
    }
}

File: app/src/main/java/com/velikececi/udfdonusturucu/core/tools/editor/UdfStructureWriter.kt
package com.velikececi.udfdonusturucu.core.tools.editor

/** UDFStructureWriter.swift'in Kotlin karşılığı — düzenleme modelini UYAP content.xml (CDATA + elements + headers + footers) biçimine yazar. */
object UdfStructureWriter {

    data class BuildResult(
        val cdataText: String,
        val elementsXML: String,
        val headersXML: String?,
        val footersXML: String?,
    )

    fun build(document: UdfEditDocument): BuildResult {
        val cdata = StringBuilder()
        val offset = IntRef(0)

        val headersXml = writeHeaderFooterSection(document.headers, itemTag = "header", cdata = cdata, offset = offset)

        val elements = StringBuilder("\n")
        for (block in document.blocks) {
            when (block) {
                is UdfEditBlock.ParagraphBlock -> elements.append(writeParagraph(block.paragraph, cdata, offset))
                is UdfEditBlock.TableBlock -> elements.append(writeTable(block.table, cdata, offset))
            }
        }

        val footersXml = writeHeaderFooterSection(document.footers, itemTag = "footer", cdata = cdata, offset = offset)

        return BuildResult(
            cdataText = cdata.toString(),
            elementsXML = elements.toString(),
            headersXML = headersXml,
            footersXML = footersXml,
        )
    }

    /** Swift'in `inout Int` ofset parametresinin Kotlin karşılığı — tek elemanlı taşıyıcı kutu. */
    private class IntRef(var value: Int)

    // MARK: - Header / Footer

    private fun writeHeaderFooterSection(items: List<UdfEditHeaderFooter>, itemTag: String, cdata: StringBuilder, offset: IntRef): String? {
        if (items.isEmpty()) return null
        val xml = StringBuilder("\n")
        for (item in items) {
            xml.append("  <$itemTag type=\"${escape(item.type)}\">\n")
            for (para in item.paragraphs) {
                xml.append(writeParagraph(para, cdata, offset, indent = "    "))
            }
            xml.append("  </$itemTag>\n")
        }
        return xml.toString()
    }

    // MARK: - Paragraph

    private fun writeParagraph(para: UdfEditParagraph, cdata: StringBuilder, offset: IntRef, indent: String = ""): String {
        var tabAttr = ""
        if (para.tabStops.isNotEmpty()) {
            tabAttr = " TabSet=\"${formatFloat(para.tabStops.first())}:0:0\""
        }

        val xml = StringBuilder(
            "$indent<paragraph SpaceAbove=\"${formatFloat(para.spaceAbove)}\" SpaceBelow=\"${formatFloat(para.spaceBelow)}\" " +
                "LeftIndent=\"${formatFloat(para.leftIndent)}\" RightIndent=\"${formatFloat(para.rightIndent)}\" " +
                "LineSpacing=\"${formatFloat(para.lineSpacing)}\" resolver=\"hvl-default\" " +
                "Alignment=\"${para.alignment}\" Hanging=\"${formatFloat(para.hangingIndent)}\"$tabAttr>",
        )

        val paraText = para.runs.joinToString("") { it.text }
        var runOffset = offset.value

        if (paraText.isEmpty()) {
            xml.append("<content resolver=\"hvl-default\" startOffset=\"${offset.value}\" length=\"0\" />")
            cdata.append("\n")
            offset.value += 1
        } else {
            for (run in para.runs) {
                if (run.text.isEmpty()) continue
                xml.append(runTag(run, startOffset = runOffset, length = run.text.length))
                runOffset += run.text.length
            }
            cdata.append(paraText).append("\n")
            offset.value += paraText.length + 1
        }

        xml.append("</paragraph>\n")
        return xml.toString()
    }

    // MARK: - Table

    private fun writeTable(table: UdfEditTable, cdata: StringBuilder, offset: IntRef): String {
        val spans = table.columnSpans.joinToString(",")
        val spansAttr = if (spans.isEmpty()) "" else " columnSpans=\"$spans\""
        val xml = StringBuilder("<table tableName=\"Tablo\" columnCount=\"${table.columnCount}\"$spansAttr border=\"${escape(table.border)}\">\n")

        for (row in table.rows) {
            xml.append("  <row rowName=\"row\" rowType=\"${escape(row.rowType)}\" border=\"${escape(table.border)}\">\n")
            for (cell in row.cells) {
                val cellAttrs = StringBuilder(
                    "colspan=\"${cell.colspan}\" rowspan=\"${cell.rowspan}\" align=\"top\" border=\"${escape(table.border)}\" borderSpec=\"15\"",
                )
                if (cell.fillColorARGB != null) {
                    cellAttrs.append(" fillColor=\"${cell.fillColorARGB}\"")
                }
                xml.append("    <cell $cellAttrs>\n")
                for (para in cell.paragraphs) {
                    xml.append(writeParagraph(para, cdata, offset, indent = "      "))
                }
                xml.append("    </cell>\n")
            }
            xml.append("  </row>\n")
        }

        xml.append("</table>\n")
        return xml.toString()
    }

    // MARK: - Run tag

    private fun runTag(run: UdfEditRun, startOffset: Int, length: Int): String {
        val tagName = when (run.kind) {
            is UdfRunKind.Content -> "content"
            is UdfRunKind.Field -> "field"
            is UdfRunKind.Space -> "space"
        }

        val attrs = StringBuilder(" resolver=\"hvl-default\"")
        if (run.isBold) attrs.append(" bold=\"true\"")
        if (run.isItalic) attrs.append(" italic=\"true\"")
        if (run.isUnderline) attrs.append(" underline=\"true\"")
        // Yalnızca varsayılan (12) dışında bir boyut varsa yazılır — iOS ile aynı, dosyayı gereksiz şişirmez.
        if (run.fontSize != 12f && run.fontSize > 0f) {
            attrs.append(" size=\"${run.fontSize.toInt()}\"")
        }
        attrs.append(" family=\"${escape(run.fontFamily)}\"")
        run.foregroundARGB?.let { attrs.append(" foreground=\"$it\"") }
        run.backgroundARGB?.let { attrs.append(" background=\"$it\"") }
        if (run.kind is UdfRunKind.Field) {
            attrs.append(" fieldName=\"${escape(run.kind.name)}\"")
        }
        attrs.append(" startOffset=\"$startOffset\" length=\"$length\"")
        return "<$tagName$attrs />"
    }

    private fun escape(value: String): String =
        value.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;")

    /** Swift'in `"\(CGFloat)"` enterpolasyonu tam sayı değerleri bile ondalıklı ("1.0") yazar; UYAP'ın kendi ayrıştırıcısıyla tutarlılık için aynısı yapılır. */
    private fun formatFloat(value: Float): String {
        val d = value.toDouble()
        return if (d == d.toLong().toDouble()) "${d.toLong()}.0" else d.toString()
    }
}
